srml/evm/src/backend.rs

```diff
@@ -0,0 +1,202 @@
+use support::storage::{StorageMap, StorageDoubleMap};
+use sha3::{Keccak256, Digest};
+use evm::Config;
+use evm::backend::{Backend as BackendT, ApplyBackend, Apply};
+use crate::{Trait, Accounts, AccountStorages, Module, Event};
+
+#[derive(Clone, Eq, PartialEq, Encode, Decode, Default)]
+#[cfg_attr(feature = "std", derive(Debug, Serialize, Deserialize))]
+/// Ethereum account nonce, balance and code. Used by storage.
+pub struct Account {
+	/// Account nonce.
+	pub nonce: U256,
+	/// Account balance.
+	pub balance: U256,
+	/// Account code.
+	pub code: Vec<u8>,
+}
+
+impl Account {
+	/// Whether the account is considered empty, in EIP-161's definition.
+	pub fn is_empty(&self) -> bool {
+		self.nonce == U256::zero() && self.balance == U256::zero() &&
+			self.code.len() == 0
+	}
+}
+
+#[derive(Clone, Eq, PartialEq, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Debug, Serialize, Deserialize))]
+/// Ethereum log. Used for `deposit_event`.
+pub struct Log {
+	/// Source address of the log.
+	pub address: H160,
+	/// Topics of the log.
+	pub topics: Vec<H256>,
+	/// Bytearray data of the log.
+	pub data: Vec<u8>,
+}
+
+#[derive(Clone, Eq, PartialEq, Encode, Decode, Default)]
+#[cfg_attr(feature = "std", derive(Debug, Serialize, Deserialize))]
+/// External input from the transaction.
+pub struct Vicinity {
+	/// Current transaction gas price.
+	pub gas_price: U256,
+	/// Origin of the transaction.
+	pub origin: H160,
+}
+
+/// Gasometer config used for executor. Currently this is hard-coded to
+/// Istanbul hard fork.
+pub static GASOMETER_CONFIG: Config = Config::istanbul();
+
+/// Substrate backend for EVM.
+pub struct Backend<'vicinity, T> {
+	vicinity: &'vicinity Vicinity,
+	_marker: PhantomData<T>,
+}
+
+impl<'vicinity, T> Backend<'vicinity, T> {
+	/// Create a new backend with given vicinity.
+	pub fn new(vicinity: &'vicinity Vicinity) -> Self {
+		Self { vicinity, _marker: PhantomData }
+	}
+
+	fn account(&self, address: H160) -> Account {
+		Accounts::get(address)
+	}
+}
+
+impl<'vicinity, T: Trait> BackendT for Backend<'vicinity, T> {
+	fn gas_price(&self) -> U256 { self.vicinity.gas_price }
+	fn origin(&self) -> H160 { self.vicinity.origin }
+
+	fn block_hash(&self, number: U256) -> H256 {
+		if number > U256::from(u32::max_value()) {
+			H256::default()
+		} else {
+			let number = T::BlockNumber::from(number.as_u32());
+			H256::from_slice(system::Module::<T>::block_hash(number).as_ref())
+		}
+	}
+
+	fn block_number(&self) -> U256 {
+		let number: u128 = system::Module::<T>::block_number().unique_saturated_into();
+		U256::from(number)
+	}
+
+	fn block_coinbase(&self) -> H160 {
+		H160::default()
+	}
+
+	fn block_timestamp(&self) -> U256 {
+		let now: u128 = timestamp::Module::<T>::get().unique_saturated_into();
+		U256::from(now)
+	}
+
+	fn block_difficulty(&self) -> U256 {
+		U256::zero()
+	}
+
+	fn block_gas_limit(&self) -> U256 {
+		U256::zero()
+	}
+
+	fn chain_id(&self) -> U256 {
+		U256::one()
```

Comment:
Can we actually pass `runtime_io::chain_id()` here?

Reply:
@pepyakin I did a quick search, looks like `runtime_io::chain_id` is hardcoded to return 42?shouldn't it be configurable in chain spec?

Reply:
That's a good question which unfortunately I cannot answer.

Reply:
#3990